cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i

```diff
@@ -0,0 +1,214 @@
+    size_t k       = par.numberOfEmbeddings;
+    size_t filtNum = par.numberOfNeighbors + 1;
+    size_t n       = xTable->getNumberOfRows(); /* Number of input feature vectors   */
+
+    SharedPtr<HomogenNumericTable<algorithmFPType> > tmpMatrixPtr =
+        HomogenNumericTable<algorithmFPType>::create(n, n, NumericTable::doAllocate, &status);
+
+    DAAL_CHECK_STATUS_VAR(status);
+    NumericTable * covOutput   = tmpMatrixPtr.get();
+    NumericTable * a0          = const_cast<NumericTable *>(xTable);
+    NumericTable * eigenvalues = const_cast<NumericTable *>(eigenTable);
+
+    // Compute cosine distances matrix
+    {
+        auto cosDistanceKernel = cosine_distance::internal::DistanceKernel<algorithmFPType, cosine_distance::Method::defaultDense, cpu>();
+        DAAL_CHECK_STATUS(status, cosDistanceKernel.compute(0, &a0, 0, &covOutput, nullptr));
+    }
+
+    WriteRows<algorithmFPType, cpu> xMatrix(covOutput, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(xMatrix);
+    algorithmFPType * x = xMatrix.get();
+
+    size_t lcnt, rcnt, cnt;
+    algorithmFPType L, R, M;
+    // Use binary search to find such d that the number of verticies having distance <= d is filtNum
+    const size_t binarySearchIterNum = 20;
+    // TODO: add parallel_for
+    for (size_t i = 0; i < n; ++i)
+    {
+        L    = 0; // min possible cos distance
+        R    = 2; // max possible cos distance
+        lcnt = 0; // number of elements with cos distance <= L
+        rcnt = n; // number of elements with cos distance <= R
+        for (size_t ij = 0; ij < binarySearchIterNum; ++ij)
+        {
+            M   = (L + R) / 2;
+            cnt = 0;
+            // Calculate the number of elements in the row with value <= M
+            for (size_t j = 0; j < n; ++j)
+            {
+                if (x[i * n + j] <= M)
+                {
+                    cnt++;
+                }
+            }
+            if (cnt < filtNum)
+            {
+                L    = M;
+                lcnt = cnt;
+            }
+            else
+            {
+                R    = M;
+                rcnt = cnt;
+            }
+            // distance threshold is found
+            if (rcnt == filtNum)
+            {
+                break;
+            }
+        }
+        // create edges for the closest neighbors
+        for (size_t j = 0; j < n; ++j)
+        {
+            if (x[i * n + j] <= R)
+            {
+                x[i * n + j] = 1.0;
+            }
+            else
+            {
+                x[i * n + j] = 0.0;
+            }
+        }
+        // fill the diagonal of matrix with zeros
+        x[i * n + i] = 0;
+    }
+
+    // Create Laplassian matrix
+    for (size_t i = 0; i < n; ++i)
+    {
+        for (size_t j = 0; j < i; ++j)
+        {
+            algorithmFPType val = (x[i * n + j] + x[j * n + i]) / 2;
+            x[i * n + j]        = -val;
+            x[j * n + i]        = -val;
```

Comment:
I'm a little bitconfused with inconsistency - in this case you are assigning values to the elements of matrix, but for the other part you are incrementing the symmetric ones. Does this matrix represent two triangular ones or should operations be actually symmetric?

Reply:
I am/was too. Hopefully a bit of clarity to help in a code comment.